wgpu-hal/src/metal/mod.rs

```diff
@@ -388,8 +390,22 @@ impl crate::Queue<Api> for Queue {
     }
 
     unsafe fn get_timestamp_period(&self) -> f32 {
-        // TODO: This is hard, see https://github.com/gpuweb/gpuweb/issues/1325
-        1.0
+        if self.shared.private_caps.supports_timestamp_period {
+            use objc::{msg_send, runtime::Object, sel, sel_impl};
+            let timestamps = [std::ptr::null_mut::<Object>(); 4];
+            let device = self.shared.device.lock().as_ptr();
+            let () = msg_send![device, sampleTimestamps:&timestamps[0] gpuTimestamp:&timestamps[1]];
```

Comment:
Maybe what we could do is grab the first sample as soon as we get the device, store it, and then take a new sample in this function so that we get the longest possible sample period to get rid of the amount of noise in the reading.

Reply:
I'm not sure if I understand it correctly, do you mean like this?
```rust
if !self.shared.private_caps.supports_timestamp_period {
   1.0
 } else {
	use objc::{msg_send, runtime::Object, sel, sel_impl};
    let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
    let device = self.shared.device.lock().as_ptr();
    let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
    if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
        1.0
    } else {
        let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
        let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
        let cpu_delta = timestamps1[0] as u64 - timestamps0[0] as u64;
        let gpu_delta = timestamps1[1] as u64 - timestamps0[1] as u64;
        cpu_delta as f32 / gpu_delta as f32
    }         
}
```

Reply:
(I think you replied to the wrong message 😄)

I meant something like

```rust
use objc::{msg_send, runtime::Object, sel, sel_impl};
if !self.shared.private_caps.supports_timestamp_period {
   return 1.0;
 } 

let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
let device = self.shared.device.lock().as_ptr();
let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
    return 1.0;
} 

let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
let cpu_delta = timestamps1[0] as u64 - timestamps0[0] as u64;
let gpu_delta = timestamps1[1] as u64 - timestamps0[1] as u64;
cpu_delta as f32 / gpu_delta as f32
```